### ISSUE TITLE
[NTI] Not Callable Warning

### ISSUE THREAD

**Issue by ChadKillingsworth** [google/closure-compiler#2310]
Using NTI with the following code (which is generated):

```js
(/** @this {window.JhaButtonElement} */ function() {
  polymerRename.symbol(12196, 12200, this.type);
  polymerRename.symbol(12217, 12225, this.disabled);
}).call(/** @type {!window.JhaButtonElement} */ (document.createElement("jha-button")))
```

I get the this warning:

```
[10:42:45] gulp-google-closure-compiler: /build/components/banno-web.template.js:3: WARNING - Cannot call non-function type undefined
(/** @this {window.JhaButtonElement} */ function() {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

However, try as I might, I can't create a simple repo case. I don't get the warning with this code:

```js
{
  /** @this {window.JhaButtonElement} */ 
  const fn = function() {
    polymerRename.symbol(12196, 12200, this.type);
    polymerRename.symbol(12217, 12225, this.disabled);
  };
  const elem = /** @type {!window.JhaButtonElement} */ (document.createElement("jha-button"));
  fn.call(elem);
}
```

**Comment by dimvar** [comment]
I can't do much without being able to repro. Can you help me repro, even with a program that is not small?

**Comment by ChadKillingsworth** [comment]
Sure thing - let me get it as small as I can and still produce the error.

**Comment by ChadKillingsworth** [comment]
@dimvar repo case emailed to you.

**Comment by dimvar** [comment]
There generated code in banno-web.template.js is missing some semicolons, and that is causing bad parsing. The warning you are seeing goes away if you put a semicolon after
`.call(/** @type {!window.JhaButtonElement} */ (document.createElement("jha-button")))`.
But then you get the same warning on the next call, and so on.

Closing since this is an issue with polymer-rename, not NTI.

**Comment by ChadKillingsworth** [comment]
Thanks - any way we can warn about ambiguous parsing like this? It only came up when I enabled NTI.

**Comment by dimvar** [comment]
Generally it's a legit pattern to call the result of a call, so I'm not sure what heuristic we would use for these. IIRC we have a heuristic to warn about object literals that are parsed as blocks. If you find a heuristic for this case that has very few false positives, we can add a warning in the parser.

As far as typing goes, I'm not sure why the old type checker didn't catch this.

**Comment by ChadKillingsworth** [comment]
It's a red flag to me anytime the call invocation is on a subsequent line:

```js
(expr)
()
```

But I'm sure someone has a style that does this and it would generate false positives.

I'm just sorry you had to look through that huge repo case just to find it was a missing semicolon.

**Comment by dimvar** [comment]
Actually the call invocation on subsequent line seems like a good heuristic to me. If you wanna give it a shot, and we'll see how many people complain :)

**Comment by ChadKillingsworth** [comment]
Well the call nodes are currently given the wrong line number by the parser. Which makes this pretty difficult.
